tests/lib/fakeinstaller/main.go

```diff
@@ -0,0 +1,282 @@
+		return nil, fmt.Errorf("cannot layout volume: %v", err)
+	}
+
+	iconst := &install.CreateOptions{CreateAllMissingPartitions: true}
+	created, err := install.CreateMissingPartitions(diskLayout, lvol, iconst)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create parititons: %v", err)
+	}
+	logger.Noticef("created %v partitions", created)
+
+	return created, nil
+}
+
+func runMntFor(label string) string {
+	return filepath.Join(dirs.GlobalRootDir, "/run/fakeinstaller-mnt/", label)
+}
+
+func postSystemsInstallSetupStorageEncryption(details *client.SystemDetails) error {
+	// TODO: check details.StorageEncryption and call POST
+	// /systems/<seed-label> with "action":"install" and
+	// "step":"setup-storage-encryption"
+	return nil
+}
+
+// XXX: reuse/extract cmd/snap/wait.go:waitMixin()
+func waitChange(chgId string) error {
+	cli := client.New(nil)
+	for {
+		chg, err := cli.Change(chgId)
+		if err != nil {
+			return err
+		}
+
+		if chg.Err != "" {
+			return errors.New(chg.Err)
+		}
+		if chg.Ready {
+			return nil
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
+
+// TODO laidoutStructs is used to get the devices, when encryption is
+// happening maybe we need to find the information differently.
+func postSystemsInstallFinish(cli *client.Client,
+	details *client.SystemDetails, bootDevice string,
+	laidoutStructs []gadget.OnDiskStructure) error {
+
+	vols := make(map[string]*gadget.Volume)
+	for volName, gadgetVol := range details.Volumes {
+		laidIdx := 0
+		for i := range gadgetVol.Structure {
+			// TODO mbr is special, what is the device for that?
+			var device string
+			if gadgetVol.Structure[i].Role == "mbr" {
+				device = bootDevice
+			} else {
+				device = laidoutStructs[laidIdx].Node
+				laidIdx++
+			}
+			gadgetVol.Structure[i].Device = device
+		}
+		vols[volName] = gadgetVol
+	}
+
+	// Finish steps does the writing of assets
+	opts := &client.InstallSystemOptions{
+		Step:      client.InstallStepFinish,
+		OnVolumes: vols,
+	}
+	chgId, err := cli.InstallSystem(details.Label, opts)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Change %s created\n", chgId)
+	return waitChange(chgId)
+}
+
+// createAndMountFilesystems creates and mounts filesystems. It returns
+// an slice with the paths where the filesystems have been mounted to.
+func createAndMountFilesystems(bootDevice string, volumes map[string]*gadget.Volume) ([]string, error) {
+	// only support a single volume for now
+	if len(volumes) != 1 {
+		return nil, fmt.Errorf("got unexpected number of volumes %v", len(volumes))
+	}
+
+	disk, err := disks.DiskFromDeviceName(bootDevice)
+	if err != nil {
+		return nil, err
+	}
+	vol := firstVol(volumes)
+
+	var mountPoints []string
+	for _, stru := range vol.Structure {
+		if stru.Label == "" || stru.Filesystem == "" {
+			continue
+		}
+
+		part, err := disk.FindMatchingPartitionWithPartLabel(stru.Label)
+		if err != nil {
+			return nil, err
+		}
+		// XXX: reuse
+		// gadget/install/content.go:mountFilesystem() instead
```

Comment:
(nitpick) this comment is probably more about the `// mount` block in line 178 ? or am I misreading this (i.e. this filesystem creation seems fine )